// @bagsfm/bags-sdk/dist/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JITO_TIP_ACCOUNTS = exports.METEORA_DAMM_V2_PROGRAM_ID = exports.METEORA_DBC_PROGRAM_ID = exports.BAGS_FEE_SHARE_V2_PROGRAM_ID = exports.BAGS_FEE_SHARE_V1_PROGRAM_ID = exports.BAGS_FEE_SHARE_V2_MAX_CLAIMERS_NON_LUT = exports.BAGS_GLOBAL_LUT = exports.BAGS_METEORA_FEE_CLAIMER_VAULT_PDA_SEED = exports.BAGS_METEORA_FEE_CLAIMER_VAULT_RENT_EXCEMPT_AMOUNT = exports.METEORA_DBC_MIGRATION_DAMM_V2_CREATOR = exports.WRAPPED_SOL_MINT = exports.BAGS_TOKEN_CREATION_AUTHORITY = exports.BAGS_PUBLIC_API_V2_DEFAULT_BASE_URL = void 0;
const idl_json_1 = __importDefault(require("./idl/dynamic-bonding-curve/idl.json"));
const idl_json_2 = __importDefault(require("./idl/damm-v2/idl.json"));
const idl_json_3 = __importDefault(require("./idl/bags-meteora-fee-claimer/idl.json"));
const idl_json_4 = __importDefault(require("./idl/fee-share-v2/idl.json"));
const web3_js_1 = require("@solana/web3.js");
exports.BAGS_PUBLIC_API_V2_DEFAULT_BASE_URL = 'https://public-api-v2.bags.fm/api/v1';
exports.BAGS_TOKEN_CREATION_AUTHORITY = new web3_js_1.PublicKey('BAGSB9TpGrZxQbEsrEznv5jXXdwyP6AXerN8aVRiAmcv');
exports.WRAPPED_SOL_MINT = new web3_js_1.PublicKey('So11111111111111111111111111111111111111112');
exports.METEORA_DBC_MIGRATION_DAMM_V2_CREATOR = new web3_js_1.PublicKey('FhVo3mqL8PW5pH5U2CN4XE33DokiyZnUwuGpH2hmHLuM');
exports.BAGS_METEORA_FEE_CLAIMER_VAULT_RENT_EXCEMPT_AMOUNT = 1398960;
exports.BAGS_METEORA_FEE_CLAIMER_VAULT_PDA_SEED = 'vault';
exports.BAGS_GLOBAL_LUT = new web3_js_1.PublicKey('Eq1EVs15EAWww1YtPTtWPzJRLPJoS6VYP9oW9SbNr3yp');
exports.BAGS_FEE_SHARE_V2_MAX_CLAIMERS_NON_LUT = 15;
exports.BAGS_FEE_SHARE_V1_PROGRAM_ID = idl_json_3.default.address;
exports.BAGS_FEE_SHARE_V2_PROGRAM_ID = idl_json_4.default.address;
exports.METEORA_DBC_PROGRAM_ID = idl_json_1.default.address;
exports.METEORA_DAMM_V2_PROGRAM_ID = idl_json_2.default.address;
exports.JITO_TIP_ACCOUNTS = [
    new web3_js_1.PublicKey('96gYZGLnJYVFmbjzopPSU6QiEV5fGqZNyN9nmNhvrZU5'),
    new web3_js_1.PublicKey('HFqU5x63VTqvQss8hp11i4wVV8bD44PvwucfZ2bU7gRe'),
    new web3_js_1.PublicKey('Cw8CFyM9FkoMi7K7Crf6HNQqf4uEMzpKw6QNghXLvLkY'),
    new web3_js_1.PublicKey('ADaUMid9yfUytqMBgopwjb2DTLSokTSzL1zt6iGPaS49'),
    new web3_js_1.PublicKey('DfXygSm4jCyNCybVYYK6DwvWqjKee8pbDmJGcLWNDXjh'),
    new web3_js_1.PublicKey('ADuUkR4vqLUMWXxW9gh6D6L8pMSawimctcNZ5pGwDcEt'),
    new web3_js_1.PublicKey('DttWaMuVvTiduZRnguLF7jNxTgiMBZ1hyAumKUiL2KRL'),
    new web3_js_1.PublicKey('3AVi9Tg9Uo68tJfuvoKvqKNWKkC5wPdSSdeBnizKZ6jT'),
];
