// @bagsfm/bags-sdk/dist/services/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseService = void 0;
const create_program_1 = require("../utils/create-program");
const bags_client_1 = require("../api/bags-client");
const state_1 = require("./state");
class BaseService {
    constructor(apiKey, connection, commitment = 'processed') {
        this.bagsApiClient = new bags_client_1.BagsApiClient(apiKey);
        this.dbcProgram = (0, create_program_1.createDbcProgram)(connection, commitment).program;
        this.dammV2Program = (0, create_program_1.createDammV2Program)(connection, commitment).program;
        this.bagsMeteoraFeeClaimer = (0, create_program_1.createBagsMeteoraFeeClaimerProgram)(connection, commitment).program;
        this.bagsFeeShareV2 = (0, create_program_1.createBagsFeeShareV2Program)(connection, commitment).program;
        this.bagsFeeShareV2Coder = (0, create_program_1.createBagsFeeShareV2Coder)();
        this.connection = connection;
        this.commitment = commitment;
        this.stateService = new state_1.StateService(apiKey, connection, commitment);
    }
}
exports.BaseService = BaseService;
